crates/orchestrator/src/jobs/da_job/mod.rs

```diff
@@ -396,25 +396,28 @@ fn da_word(class_flag: bool, nonce_change: Option<Felt>, num_changes: u64) -> Fe
 }
 
 fn refactor_state_update(state_update: &mut StateDiff) {
-    let addresses_in_nonces: Vec<Felt> = state_update.nonces.clone().iter().map(|item| item.contract_address).collect();
-    let addresses_in_storage_diffs: Vec<Felt> =
-        state_update.storage_diffs.clone().iter().map(|item| item.address).collect();
-
-    let address_to_insert = find_unique_addresses(addresses_in_nonces, addresses_in_storage_diffs);
-
-    for address in address_to_insert {
-        state_update.storage_diffs.push(ContractStorageDiffItem { address, storage_entries: vec![] })
-    }
-}
-
-fn find_unique_addresses(nonce_addresses: Vec<Felt>, storage_diff_addresses: Vec<Felt>) -> Vec<Felt> {
-    let storage_set: HashSet<_> = storage_diff_addresses.into_iter().collect();
-
-    nonce_addresses.into_iter().filter(|addr| !storage_set.contains(addr)).collect()
+    let existing_storage: HashSet<_> = state_update.storage_diffs.iter().map(|item| item.address).collect();
+
+    let required_addresses: HashSet<_> = Iterator::chain(
+        state_update.nonces.iter().map(|item| item.contract_address),
+        Iterator::chain(
+            state_update.deployed_contracts.iter().map(|item| item.address),
+            existing_storage.iter().copied(),
+        ),
+    )
+    .collect();
+
+    // Add missing storage diffs in one batch
+    state_update.storage_diffs.extend(
+        required_addresses
+            .difference(&existing_storage)
+            .map(|&address| ContractStorageDiffItem { address, storage_entries: Vec::new() }),
+    );
```

Comment:
ok this is definitely cleaner than before but can you check how many clones this is doing? from my understanding we're copying addresses into `existing_storage` and then into `required_addresses`. Instead, we could 

1. loop over an iterator chain of nonces + deployed contracts
2. if not present in `existing_storage`, insert into `existing_storage` and `state_update.storage_diffs`

This way, I think we will have 
1. A smaller loop (currently we loop on `required_addresses` inside `difference` I beleive and required addresses has storage_diffs, nonce and deployed contracts. now we only loop on nonce and deployed contracts.
2. Less memory, you don't need to have `required_addresses`

Reply:
what do you think about this approach ?  

```rust
fn refactor_state_update(state_update: &mut StateDiff) {
    let existing_storage: HashSet<_> = state_update.storage_diffs.iter().map(|item| item.address).collect();

    // Collect new addresses, using a HashSet for deduplication
    let new_addresses: HashSet<_> = Iterator::chain(
        state_update.nonces.iter().map(|item| item.contract_address),
        state_update.deployed_contracts.iter().map(|item| item.address),
    )
    .filter(|address| !existing_storage.contains(address))
    .collect();

    // Add new storage diffs in batch
    state_update.storage_diffs.extend(
        new_addresses.into_iter().map(|address| ContractStorageDiffItem { address, storage_entries: Vec::new() }),
    );
}
```

Reply:
looks good!